pkg/controller/servicemesh/memberroll/controller.go

```diff
@@ -598,6 +636,126 @@ func (r *defaultKialiReconciler) reconcileKiali(ctx context.Context, kialiCRName
 	return nil
 }
 
+type PrometheusReconciler interface {
+	reconcilePrometheus(ctx context.Context, prometheusCMName, prometheusNamespace string, configuredMembers []string) error
+}
+
+type defaultPrometheusReconciler struct {
+	Client client.Client
+}
+
+func (r *defaultPrometheusReconciler) reconcilePrometheus(ctx context.Context, prometheusCMName, prometheusNamespace string, members []string) error {
+	reqLogger := common.LogFromContext(ctx)
+	reqLogger.Info("Attempting to get Prometheus ConfigMap", "prometheusNamespace", prometheusNamespace, "prometheusCMName", prometheusCMName)
+
+	cm := &corev1.ConfigMap{}
+
+	err := r.Client.Get(ctx, client.ObjectKey{Name: prometheusCMName, Namespace: prometheusNamespace}, cm)
+	if err != nil {
+		if meta.IsNoMatchError(err) || errors.IsNotFound(err) || errors.IsGone(err) {
+			reqLogger.Info("Prometheus Config Map does not exist.")
+			return nil
+		}
+		return pkgerrors.Wrap(err, "error retrieving Prometheus ConfigMap from mesh")
+	}
+
+	data := make(map[string]interface{})
+
+	err = yaml.Unmarshal([]byte(cm.Data[prometheusConfigurationFilename]), &data)
+
+	if err != nil {
+		return pkgerrors.Wrap(err, "error unmarshaling prometheus.yml")
+	}
+
+	f, ok, err := unstructured.NestedFieldNoCopy(data, prometheusScrapeConfigKeyName)
+
+	var scrapes []interface{}
+
+	if err == nil && ok {
+		scrapes, ok = f.([]interface{})
+
+		if !ok {
+			return fmt.Errorf("error no scrape_configs found from %v", f)
+		}
+	} else {
+		return pkgerrors.Wrap(err, "error getting scrape_configs from ConfigMap")
+	}
+
+	for _, s := range scrapes {
+		scrape, ok := s.(map[string]interface{})
+
+		if !ok {
+			return fmt.Errorf("error converting scrape_config from %v", s)
+		}
+
+		if scrape["job_name"] == "pilot" {
+			continue
+		}
+
+		f, ok, err := unstructured.NestedFieldNoCopy(scrape, prometheusScrapeSDConfigKeyName)
+
+		if err == nil {
+			if ok {
+				sds, ok := f.([]interface{})
+
+				if ok {
+					for _, v := range sds {
+						sd, ok := v.(map[string]interface{})
+
+						if ok {
+							reqLogger.Info(fmt.Sprintf("Updating sd %v", v))
+
+							err = unstructured.SetNestedStringSlice(sd, members, "namespaces", "names")
+
+							if err != nil {
+								return pkgerrors.Wrap(err, fmt.Sprintf("error setting sd %v", v))
+							}
+						}
+					}
+				} else {
+					return fmt.Errorf("error can not process sd %v", f)
+				}
+			} else {
+				reqLogger.Info(fmt.Sprintf("Ignoring scrape %v", s))
+			}
+		} else {
+			return pkgerrors.Wrap(err, fmt.Sprintf("error getting sd from %v", s))
+		}
+	}
```

Comment:
The logic above is quite tricky and difficult to read, so I think that we have to extract it to a separate function and write unit tests to make sure this is correct.

Reply:
+1 and let's move this into `prometheus.go` (and maybe the kiali stuff to `kiali.go` while we're at it)